### PR TITLE
Refactor comment api

### PR DESCRIPTION
**Pull request by Pycea**
Improves separation of options and comments.

1. Removes dependency of `ExtComment(List)Component` on option api. Comments are now completely separate from options, and option functions are called on comments externally. Also removes the requirement of the option value to be truthy in order for `processComment` to run, instead passing the current value to the function.

2. Separates out functions to create the reply/edit/delete buttons. Perhaps more controversially, removes the `connectCommentEditor` and `enableCommentReply` functions. While this does lead to a bit more redundancy, the nested callback structure and numerous params seem to cost more clarity than they're worth. I'd also argue that while replying to a comment and creating a top level comment happen to do similar things, they're not an inherent logical grouping.

    Instead, some of the redundant functionality, like hiding and showing elements and removing the edit box, has been moved to the `CommentEditorComponent` class.

### CODE REVIEW

File: extension/ext-comments.js [hunk 1]
```diff
@@ -211,49 +212,14 @@ class ExtCommentListComponent {
     for (const child of this.children) child.reverse();
   }
 
-  // Applies the given list of options to all descendant comments.
-  applyOptions(options) {
-    if (options.length === 0) return;
-    for (const child of this.children) child.applyOptions(options);
+  *allChildren() {
```
Comment (maksverver) [inline extension/ext-comments.js]: Please rename this to `descendants`.

File: extension/options.js [hunk 2]
```diff
@@ -293,11 +313,13 @@ const {
   // Keep this list in sync with the object at the top of the file.
   return {
     OPTIONS,
+    processComments,
```
Comment (maksverver) [inline extension/options.js]: I think `processComments()` doesn't need to be exported?

(If you remove it here, remember to also remove it at the top of the file.)

File: extension/options.js [hunk 3]
```diff
@@ -139,12 +142,12 @@ const {
     },
     onValueChange(newValue) {
       this.createCachedSet(newValue);
-      reprocessComments(this);
+      processComments(this.key);
```
Comment (maksverver) [inline extension/options.js]: Is this change needed? It seems like the first thing processComments() does is look up the option by key, so why not pass `this` directly as before?

(You also need to update `processAllComments()` accordingly, but that's also trivial.)

File: extension/ext-comments.js [hunk 4]
```diff
@@ -663,7 +645,12 @@ class CommentOrderComponent {
 }
 
 class CommentEditorComponent {
-  constructor(parentElem, initialText, callback) {
+  static SUCCESS = 0;
+  static ERROR = 1;
```
Comment (maksverver) [inline extension/ext-comments.js]: Comment: I'm not a fan of these constants here.

If the idea is that the `exitCallback` should be able to prevent the dialog from closing, then I'd prefer either returning a simple boolean, or throwing an exception to indicate failure.

If exitCallback throws an exception, we could unify the error handling here:

   console.warn(e);
   alert('Failed to post comment!\n\nSee the JavaScript console for details.');
   
Which removes a little duplication, which might be nice!

Anyway, this is not a blocker for me, but I might refactor this later.

**Review by maksverver — CHANGES_REQUESTED** [review]

File: extension/options.js (hunk 3 above)
Comment (Pycea) [reply extension/options.js]: You're probably right. The idea was that other things could call it by only needing the key, but it isn't exported anymore and they have access to OPTIONS either way.

File: extension/ext-comments.js (hunk 4 above)
Comment (Pycea) [reply extension/ext-comments.js]: This is definitely the way.
